[ENH] vectorization/iteration utility for `sktime` time series formats

fkiraly opened the pull request:
This PR introduces a vectorization/iteration utility for the two use cases:
* a `Panel` is passed to a series transformer, which then loops through all the series in the panel. That is, `fit` and `transform` are applied iteratively to all series, the result is then combined back to a panel.
* a `Panel` is passed to a simple forecaster, which loops through all the series in a simple panel or a hierarchical panel. That is, `fit` and `predict` are applied iteratively, and the result is combained back to a panel forecast.

This utility would facilitate to use all current forecasters and transformers in an automatically vectorized fashion, where `fit` etc are automatically cast to the dimensions defined in the panel and hierarchical indices.

The current transformer base class refactor in #1790 does not yet use this utility, but could be easily refactored.
The extension to hierarchical forecasting would also profit from this, the original motivation to have this is that this would constitute a second use case. In consequence, this PR also makes use of PR #1786 (the experimental hierarchical time series scitype).
This is also dependent on #1805 which was uncovered through testing.

In detail, content introduced in this PR:
* the module `datatypes._vectorize` which contains the class `VectorizedDF`. It enables easy looping through slices of `Series` and `Panel` data which are automatically converted to `sktime` formatting conventions of `pandas.DataFrame` based containers.
* a series of tests for `VectorizedDF`, in `datatypes.test.test_vectorize`

fkiraly commented:
FYI @danbartl, @ltsaprounis, @satya-pattnaik, this would be useful for the "auto-vectorization" functionality in hierarchical forecasters. I've factored it out in line with the principle "keep PR small".

chrisholder commented on sktime/datatypes/_hierarchical/_check.py:
```diff
@@ -0,0 +1,127 @@
+        return valid
+
+
+def check_pdmultiindex_Hierarchical(obj, return_metadata=False, var_name="obj"):
```
Why is Hierarchical capitalised

fkiraly replied:
this is from the dependency #1786, was addressed there.

chrisholder commented on sktime/datatypes/_hierarchical/_convert.py:
```diff
@@ -0,0 +1,25 @@
+
+def convert_identity(obj, store=None):
+
+    return obj
```
Im assuming this intended to be expanded upon in the future?

fkiraly replied:
this is from the dependency #1786, was addressed there.

chrisholder commented on sktime/datatypes/_vectorize.py:
```diff
@@ -0,0 +1,209 @@
+# -*- coding: utf-8 -*-
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
```
Since when have we had a copyright and is this now needed on all files?

fkiraly replied:
I thought we always had this? Just inconsistently added.

chrisholder commented on sktime/datatypes/_vectorize.py:
```diff
@@ -0,0 +1,209 @@
+
+    Parameters
+    ----------
+    X : Panel or Hierarchical data container, to vectorize over
```
Type and description need to seperated by - or new line

fkiraly replied:
fixed

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+def _generate_scitype_mtype_combinations():
+    """Return scitype/mtype tuples for pytest_generate_tests.
+
+    Fixtures parameterized
```
Why is this called Fixtures parameterized? Should just be Parameters

fkiraly replied:
well, it's not parameters, since these are no function parameters.
Not sure how to describe it, since it produces fixture combinations.
Can also remove it?

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+def _generate_scitype_mtype_fixtureindex_combinations():
+    """Return fixture tuples for pytest_generate_tests.
+
+    Fixtures parameterized
```
Same

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+def pytest_generate_tests(metafunc):
+    """Test parameterization routine for pytest.
+
+    Fixtures parameterized
```
Same

fkiraly replied:
`pytest_generate_tests` is a way to parameterize fixtures that cannot be easily written as cartesian products.
The fixtures that this parameterizes are listed, so people reading the code have access to the information easily.

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+def test_construct_VectorizedDF(scitype, mtype, fixture_index):
+    """Test VectorizedDF constructs with valid arguments.
+
+    Fixtures parameterized
```
Same

fkiraly replied:
Here the fixtures are being used.

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+def test_construct_VectorizedDF_errors(scitype, mtype, fixture_index):
+    """Test VectorizedDF raises appropriate errors with invalid arguments.
+
+    Fixtures parameterized
```
Same

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+def test_item_len(scitype, mtype, fixture_index, iterate_as):
+    """Tests __len__ returns correct length.
+
+    Fixtures parameterized
```
Same

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+        VectorizedDF(X=fixture, iterate_as="Panel", is_scitype=None)
+
+
+def test_construct_VectorizedDF_errors(scitype, mtype, fixture_index):
```
VecotrizedDf shouldnt be capatilized

fkiraly replied:
well, but it's the name of the class that is tested. Should I still lower-case it?

chrisholder replied:
Yea always lower case in function names even refering to a class

fkiraly replied:
ok, will change

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+        metafunc.parametrize("iterate_as", ["Panel", "Series"])
+
+
+def test_construct_VectorizedDF(scitype, mtype, fixture_index):
```
VecotrizedDf shouldnt be capatalized

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+def test_iteration(scitype, mtype, fixture_index, iterate_as):
+    """Tests __getitem__ returns pd-multiindex mtype if iterate_as="Series".
+
+    Fixtures parameterized
```
Same

chrisholder commented on sktime/datatypes/tests/test_vectorize.py:
```diff
@@ -0,0 +1,321 @@
+def test_series_item_mtype(scitype, mtype, fixture_index, iterate_as):
+    """Tests __getitem__ returns correct pd-multiindex mtype.
+
+    Fixtures parameterized
```
Same

fkiraly commented:
@chrisholder, I've addressed some points and commented on others.
* the "fixtures parameterized" section is similar to the same section in `test_all_estimators`. If you don't like it, we need to change the template overall.
* the copyright notice, I thought, was to be included in all files? At least it doesn't hurt since it refers to the main LICENSE.

fkiraly commented:
@chrisholder, FYI, I fixed the outstanding capitalization (lower_snake_case) and docstring issues.

I haven't changed the "fixtures parameterized" section, since that's similar to `test_all_estimators`.

fkiraly commented:
> Indeed! I made something similar (a bit hacky) a while back to benchmark global models against univariate model

Yes, exactly - I tried to turn the general principle into cleanly designed code.
